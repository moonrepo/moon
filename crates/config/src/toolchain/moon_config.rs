use crate::config_struct;
use schematic::{Config, validate};

config_struct!(
    /// Configures how and where updates will be received.
    #[derive(Config)]
    pub struct MoonConfig {
        /// A secure URL to lookup the latest version.
        #[setting(validate = validate::url_secure, default = "https://launch.moonrepo.app/versions/cli/current")]
        pub manifest_url: String,

        /// A secure URL for downloading the moon binary.
        #[setting(validate = validate::url_secure, default = "https://github.com/moonrepo/moon/releases/latest/download")]
        pub download_url: String,
    }
);
