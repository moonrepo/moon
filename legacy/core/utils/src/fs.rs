use crate::time::{is_stale, to_millis};
use moon_common::consts::CONFIG_DIRNAME;
use starbase_utils::fs;
use std::path::{Path, PathBuf};
use std::time::Duration;

pub mod temp {
    use super::*;

    pub fn get_dir() -> PathBuf {
        crate::get_workspace_root()
            .join(CONFIG_DIRNAME)
            .join("cache")
            .join("temp")
    }

    pub fn get_file(source: &str, ext: &str) -> PathBuf {
        get_dir().join(format!("{}.{}", crate::hash(source), ext))
    }

    pub fn read<P: AsRef<Path>>(path: P) -> miette::Result<Option<String>> {
        let file = path.as_ref();

        if !file.exists() {
            return Ok(None);
        }

        if let Ok(metadata) = file.metadata()
            && let Ok(filetime) = metadata.created()
        {
            // Temp files only last for 4 hours (half a workday)
            if is_stale(to_millis(filetime), Duration::from_secs(60 * 60 * 4)) {
                fs::remove_file(file)?;

                return Ok(None);
            }
        }

        Ok(Some(fs::read_file(file)?))
    }

    pub fn write<P: AsRef<Path>, D: AsRef<str>>(path: P, data: D) -> miette::Result<()> {
        let path = path.as_ref();

        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        fs::write_file(path, data.as_ref())?;

        Ok(())
    }
}
