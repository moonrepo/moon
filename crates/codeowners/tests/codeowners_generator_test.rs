use moon_codeowners::CodeownersGenerator;
use moon_config::{ConfigLoader, VcsProvider};
use starbase_sandbox::{Sandbox, assert_snapshot, create_empty_sandbox, locate_fixture};
use std::fs;

fn load_generator(provider: VcsProvider) -> Sandbox {
    let sandbox = create_empty_sandbox();
    let config_loader = ConfigLoader::default();

    sandbox.create_file(
        ".moon/workspace.yml",
        fs::read_to_string(locate_fixture("workspace").join("workspace.yml")).unwrap(),
    );

    let mut generator = CodeownersGenerator::new(sandbox.path(), provider).unwrap();
    let workspace_config = config_loader.load_workspace_config(sandbox.path()).unwrap();

    generator
        .add_workspace_entries(&workspace_config.codeowners)
        .unwrap();

    for project_fixture in ["custom-groups", "list-paths", "map-paths", "no-paths"] {
        sandbox.create_file(
            format!("{}/moon.yml", project_fixture),
            fs::read_to_string(locate_fixture(project_fixture).join("moon.yml")).unwrap(),
        );

        let project_config = config_loader
            .load_project_config_from_source(sandbox.path(), project_fixture)
            .unwrap();

        generator
            .add_project_entry(
                project_fixture,
                project_fixture,
                &project_config.owners,
                &workspace_config.codeowners,
            )
            .unwrap();
    }

    generator.generate().unwrap();

    sandbox
}

mod codeowners {
    use super::*;

    #[test]
    fn generates_bitbucket() {
        let sandbox = load_generator(VcsProvider::Bitbucket);

        assert_snapshot!(fs::read_to_string(sandbox.path().join("CODEOWNERS")).unwrap());
    }

    #[test]
    fn generates_github() {
        let sandbox = load_generator(VcsProvider::GitHub);

        assert_snapshot!(fs::read_to_string(sandbox.path().join(".github/CODEOWNERS")).unwrap());
    }

    #[test]
    fn generates_gitlab() {
        let sandbox = load_generator(VcsProvider::GitLab);

        assert_snapshot!(fs::read_to_string(sandbox.path().join(".gitlab/CODEOWNERS")).unwrap());
    }

    #[test]
    fn generates_other() {
        let sandbox = load_generator(VcsProvider::Other);

        assert_snapshot!(fs::read_to_string(sandbox.path().join("CODEOWNERS")).unwrap());
    }
}
