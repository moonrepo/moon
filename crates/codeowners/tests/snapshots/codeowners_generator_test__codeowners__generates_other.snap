---
source: crates/codeowners/tests/codeowners_generator_test.rs
expression: "fs::read_to_string(sandbox.path().join(\"CODEOWNERS\")).unwrap()"
---
# Automatically generated by moon. DO NOT MODIFY!
# https://moonrepo.dev/docs/guides/codeowners

# (workspace)
* @admins
/config @infra
/*.js @infra @frontend

# custom-groups
/custom-groups/path/with\ space/ @@@group

# list-paths
/list-paths/*.{js,ts,tsx} @frontend
/list-paths/*.config.js @frontend

# map-paths
/map-paths/*.rs @rust-dev @api-team
/map-paths/config/ @backend
